generator/generator.go

```diff
@@ -64,29 +66,38 @@ func (g *generator) BatchOperations(logger lager.Logger) (map[string]operationq.
 	evacuatingLRPs := make(map[string]models.ActualLRP)
 	tasks := make(map[string]*models.Task)
 
-	errChan := make(chan error, 3)
-
 	logger.Info("getting-containers-lrps-and-tasks")
+
+	var wg sync.WaitGroup
+	wg.Add(3)
+	var groupErr error
+
 	go func() {
+		defer wg.Done()
+
 		foundContainers, err := g.executorClient.ListContainers(logger)
 		if err != nil {
 			logger.Error("failed-to-list-containers", err)
 			err = fmt.Errorf("failed to list containers: %s", err.Error())
+			groupErr = multierror.Append(groupErr, err)
```

Comment:
The fundamental data race we observed earlier is still present, and in some sense is worse, as each of the three goroutines is doing a read-modify-write operation on the shared `groupErr` memory location. In keeping with the Golang slogan of ["share memory by communicating"](https://golang.org/doc/effective_go.html#sharing), maybe it would make sense still to have the goroutines pass their errors back to the goroutine that invoked them via a channel, and for that goroutine to accumulate them into a multierror.

Agreed that the earlier return would be correct, though: if the caller gets an error from the callee, it should assume that the other return values are invalid and should not process them, even if we happen to know that currently only a `nil` or an empty slice would be returned there.

Reply:
:+1: to both switching back to using channels and using multi-error

Reply:
reverted this, and only use multi error now.